examples/source_separation/train.py

```diff
@@ -0,0 +1,169 @@
+    )
+    group.add_argument(
+        "--num-workers",
+        type=int,
+        default=max_world_size,
+        help=(
+            "The size of distributed trainig workers. "
+            "If launching a training as single-node, multi-worker training, "
+            "(i.e. --worker-id is not provided) then this value should not exceed "
+            "the number of available GPUs. "
+            "If launching the training process as a multi-node, multi-gpu training, "
+            "(i.e. --worker-id is provided) then the value has to match "
+            "the number of workers across nodes."
+        ),
+    )
+    group.add_argument(
+        "--sync-protocol",
+        type=str,
+        default="env://",
+        help=(
+            "Synchronization protocol for distributed training. "
+            "This value is passed as ``init_method`` argument of "
+            "``torch.distributed.init_process_group`` function."
+            'If you are using ``"env://"``, you can additionally configure '
+            'environment variables "MASTER_ADDR" and "MASTER_PORT". '
+            'If you are using ``"file://..."``, then the process has to have '
+            "the access to the designated file. "
+            "See the documentation for ``torch.distributed`` for the detail. "
+            'If you are running the training in a single node, ``"env://"`` '
+            "should do. If you are running the training in multiple nodes, "
+            "you need to provide the file location where all the nodes have "
+            'access, using ``"file://..."`` protocol.'
+        ),
+    )
+    parser.add_argument(
+        "rest", nargs=argparse.REMAINDER, help="Model-specific arguments."
+    )
+    namespace = parser.parse_args(args)
+    if namespace.worker_id is None:
+        if namespace.device_id is not None:
+            raise ValueError(
+                "`--device-id` cannot be provided when runing as master process."
+            )
+        if namespace.num_workers > max_world_size:
+            raise ValueError(
+                "--num-workers ({num_workers}) cannot exceed {device_count}."
+            )
+    if namespace.rest[:1] == ["--"]:
+        namespace.rest = namespace.rest[1:]
+    return namespace
+
+
+def _main(cli_args):
+    args = _parse_args(cli_args)
+
+    if any(arg in ["--help", "-h"] for arg in args.rest):
+        _run_training(args.rest)
+
+    _init_logger(args.worker_id, args.debug)
+    if args.worker_id is None:
+        _run_training_subprocesses(args.num_workers, cli_args)
+    else:
+        dist_utils.setup_distributed(
+            world_size=args.num_workers,
+            rank=args.worker_id,
+            local_rank=args.device_id,
+            init_method=args.sync_protocol,
+        )
+        if torch.cuda.is_available():
+            torch.cuda.set_device(args.device_id)
+            _LG.info("CUDA device set to %s", args.device_id)
+        _run_training(args.rest)
```

Comment:
I had a better experience (much faster execution for some reason) when using `torch.distributed.launch` instead of spawning the jobs with `torch.multiprocessing.spawn`. This might be up to users' preference, but I generally let slurm spawn the processes for me, I only need to handle single process in my scripts.

Reply:
When running from SLURM, this script does not spawn any subprocess. The main process serves as the training worker, so  neither of `torch.distributed.launch` or `torch.multiprocessing.spawn` is used here.